crates/wasm/src/storage.rs

```diff
@@ -0,0 +1,179 @@
+                id.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_note(
+        &self,
+        commitment: &note::StateCommitment,
+    ) -> WasmResult<Option<SpendableNoteRecord>> {
+        let tx = self.db.transaction_on_one("spendable_notes")?;
+        let store = tx.object_store("spendable_notes")?;
+
+        store
+            .get_owned(base64::Engine::encode(
+                &base64::engine::general_purpose::STANDARD,
+                commitment.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_note_by_nullifier(
+        &self,
+        nullifier: &Nullifier,
+    ) -> WasmResult<Option<SpendableNoteRecord>> {
+        let tx = self.db.transaction_on_one("spendable_notes")?;
+        let store = tx.object_store("spendable_notes")?;
+
+        store
+            .index("nullifier")?
+            .get_owned(&base64::Engine::encode(
+                &base64::engine::general_purpose::STANDARD,
+                nullifier.to_proto().inner,
+            ))?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn store_advice(&self, note: Note) -> WasmResult<()> {
+        let tx = self.db.transaction_on_one("notes")?;
+        let store = tx.object_store("notes")?;
+
+        let note_proto: penumbra_proto::core::crypto::v1alpha1::Note = note.clone().try_into()?;
+        let note_js = serde_wasm_bindgen::to_value(&note_proto)?;
+
+        let commitment_proto = note.commit().to_proto();
+
+        let commitment_js = serde_wasm_bindgen::to_value(&commitment_proto)?;
+
+        store.put_key_val_owned(commitment_js, &note_js)?;
+
+        Ok(())
+    }
+
+    pub async fn read_advice(&self, commitment: note::StateCommitment) -> WasmResult<Option<Note>> {
+        let tx = self.db.transaction_on_one("notes")?;
+        let store = tx.object_store("notes")?;
+
+        let commitment_proto = commitment.to_proto();
+
+        let commitment_js = serde_wasm_bindgen::to_value(&commitment_proto)?;
+
+        store
+            .get_owned(commitment_js)?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
+    }
+
+    pub async fn get_chain_parameters(&self) -> WasmResult<Option<ChainParameters>> {
+        let tx = self.db.transaction_on_one("chain_parameters")?;
+        let store = tx.object_store("chain_parameters")?;
+
+        store
+            .get_owned("chain_parameters")?
+            .await?
+            .map(serde_wasm_bindgen::from_value)
+            .transpose()
+            .map_err(WasmError::from)
```

Comment:
With `thiserror`, you shouldn't need to be doing `.map_err(WasmError::from)` anymore. Just put the error in the new Error struct.